drivers/azure/azure.go

```diff
@@ -0,0 +1,455 @@
+package azure
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+
+	"github.com/docker/machine/drivers/azure/azureutil"
+	"github.com/docker/machine/libmachine/drivers"
+	"github.com/docker/machine/libmachine/log"
+	"github.com/docker/machine/libmachine/mcnflag"
+	"github.com/docker/machine/libmachine/state"
+
+	"github.com/Azure/azure-sdk-for-go/arm/storage"
+)
+
+const (
+	defaultAzureEnvironment     = "AzurePublicCloud"
+	defaultAzureResourceGroup   = "docker-machine"
+	defaultAzureSize            = "Standard_A2"
+	defaultAzureLocation        = "westus"
+	defaultSSHUser              = "ubuntu" // 'root' not allowed on Azure
+	defaultDockerPort           = 2376
+	defaultAzureImage           = "canonical:UbuntuServer:14.04.3-LTS:latest"
+	defaultAzureVNet            = "docker-machine-vnet"
+	defaultAzureSubnet          = "docker-machine"
+	defaultAzureSubnetPrefix    = "192.168.0.0/16"
+	defaultStorageType          = storage.StandardLRS
+	defaultAzureAvailabilitySet = "docker-machine"
+)
+
+const (
+	flAzureEnvironment     = "azure-environment"
+	flAzureSubscriptionID  = "azure-subscription-id"
+	flAzureResourceGroup   = "azure-resource-group"
+	flAzureSSHUser         = "azure-ssh-user"
+	flAzureDockerPort      = "azure-docker-port"
+	flAzureLocation        = "azure-location"
+	flAzureSize            = "azure-size"
+	flAzureImage           = "azure-image"
+	flAzureVNet            = "azure-vnet"
+	flAzureSubnet          = "azure-subnet"
+	flAzureSubnetPrefix    = "azure-subnet-prefix"
+	flAzureAvailabilitySet = "azure-availability-set"
+	flAzurePorts           = "azure-open-port"
+	flAzurePrivateIPAddr   = "azure-private-ip-address"
+	flAzureUsePrivateIP    = "azure-use-private-ip"
+	flAzureNoPublicIP      = "azure-no-public-ip"
+)
+
+const (
+	driverName = "azure"
+	sshPort    = 22
+)
+
+// Driver represents Azure Docker Machine Driver.
+type Driver struct {
+	*drivers.BaseDriver
+
+	Environment    string
+	SubscriptionID string
+	ResourceGroup  string
+
+	DockerPort      int
+	Location        string
+	Size            string
+	Image           string
+	VirtualNetwork  string
+	SubnetName      string
+	SubnetPrefix    string
+	AvailabilitySet string
+
+	OpenPorts     []string
+	PrivateIPAddr string
+	UsePrivateIP  bool
+	NoPublicIP    bool
+
+	// Ephemeral fields
+	ctx        *azureutil.DeploymentContext
+	resolvedIP string // cache
+}
+
+// NewDriver returns a new driver instance.
+func NewDriver(hostName, storePath string) drivers.Driver {
+	// NOTE(ahmetalpbalkan): any driver initialization I do here gets lost
+	// afterwards, especially for non-Create RPC calls. Therefore I am mostly
+	// making rest of the driver stateless by just relying on the following
+	// piece of info.
```

Comment:
@nathanleclaire I believe from what I tried, assigning certain objects during various points of this object's lifetime caused those instances to get lost. So it is not very clear to me at which exact point this config data gets marshaled and unmarshaled.

For instance, when I assign something here during create or pre-create, I saw that it totally gets lost during GetState/GetIP calls that occur later during the machine creation.

For quite a moment I thought maybe it is reinitializing the driver for those calls –but this is not an external driver, I was assuming it'd be in-process and there'd be only a single azure Driver instance throughout the lifetime of a docker-machine execution process; but it appears like due to RPC that's not the case.

I'm still not clear about inner workings of RPC model in docker-machine. I  found my way around it but guess I still appreciate any explanation about lifecycle of Driver instance and when/how config gets (un/)marshaled.